### DISCUSSION TITLE
Http headers aren't applied to site root, only pages

### DISCUSSION THREAD

**Discussion by paulrudy** [vercel/community#2638] [Help] [answered]
### Summary

In `vercel.json`, I can't get headers to apply universally to every page including the site root. In `vercel dev`, they are, but after deployment, the headers aren't applied to the site root.

Any suggestions would be appreciated!

### Example

[Simple example using Next.js starter](https://nextjs-test-zeta-ashy.vercel.app)
[Github repo for example](https://github.com/paulrudy/nextjs-test)

### Steps to Reproduce

In a project such as the Next.js starter project create `vercel.json` and specify some headers:

```json
{
  "headers": [
    {
      "source": "/(.*)",
      "headers": [
        {
          "key": "X-Content-Type-Options",
          "value": "nosniff"
        },
        {
          "key": "X-Frame-Options",
          "value": "DENY"
        }
      ]
    }
  ]
}
```

In `vercel dev`, the headers show up on the site root (e.g., https://mysite.com) as well as any other pages (e.g., https://mysite.com/page-1).

But after deploying to Vercel, the headers are only applied to pages _other than_ the site root.

I've tried replacing the `source` value with `/:path*`, but I get the same result.

**Comment by z4nr34l** [comment]
If you're going to use Next.js in this project maybe this would be a better way to send headers from framework itself, so they should be defined in `next.config.js` as follows:

```
/** @type {import("next").NextConfig} */
const nextConfig = {
  headers() {
    return [
      {
        source: "/(.*)",
        headers: [
          {
            "key": "X-Content-Type-Options",
            "value": "nosniff"
          },
          {
            "key": "X-Frame-Options",
            "value": "DENY"
          }
        ]
      }
    ];
  }
}

module.exports = nextConfig;
```

Which is a field-tested solution with such problem. If it helped please consider marking my solution as answer so others would be able to find that solution quickly. If you will need further assistance in this case feel free to ping me so I would be able to help! :)

**Reply by paulrudy** [reply]
Thanks, although I only used Next.js to provide a simple demonstration. I'm not actually using it in my project.

What I'm looking for is an explanation for why the source value as given doesn't work, or what I can replace it with that _will_ work.

**Reply by z4nr34l** [reply]
Ok, that's changing case a bit :D I will try to look into it and figure out solution for vercel deployment solution ;)

**Reply by z4nr34l** [reply]
I've made test deployment for you, and here are the results (as everything just works fine).

[Deployment URL](https://next-vercel-headers.vercel.app/)
[Deployment Repository](https://github.com/z4nr34l/next-vercel-headers)

As you can see I was not able to reproduce your problem using informations that you provided.

If that would be easier for you feel free to clone my example and test it on your own.

If only you will need help with your project in other framework, please provide me more details so I would be able to try to reproduce this issue on my own.

**Reply by paulrudy** [reply]
Thanks @z4nr34l, I appreciate you taking the time to check it out. I do see that there's no issue on your example. Any idea why it's working in your example and not in mine? I posted the github repo for my example in my original post.

**Reply by paulrudy** [reply]
Actually, upon closer examination @z4nr34l , your root page does not receive the specified headers some of the time. In Chrome, the headers from `vercel.json` aren't there on first load or after each refresh of the page. They do appear after visiting "subpage" and then returning home. In Safari, the specified headers always appear on the home page.

**Reply by z4nr34l** [reply, marked as answer]
@paulrudy The case that you described now is not a bug, please take a look at `HTTP Reponse Code` in requests you are checking. Requests with code `304 Not Modified` will not have those headers, as those headers were not changed too.

First request should always have code `200 OK` and headers in it, later when you refresh if nothing changed you get `304 Not Modified`.

I hope I have succeeded in showing in a simple way how the mechanism works :D

When you want to debug response headers in Chrome, use `CMD + SHIFT + R` / `CTRL + SHIFT + R` to reload without cache so chrome will not give you `304 Not Modified`.

Also please copy my `vercel.json` to your repository, redeploy and check if it works now :)

**Reply by paulrudy** [reply]
Thanks for your patience and for the information about `304 Not Modified`. I think this was creating the illusion for me that the headers weren't being applied. All set!

**Reply by z4nr34l** [reply]
I'm happy that I was able to help! Have a nice time coding!
